lib/src/Pos/Diffusion/Full/Block.hs

```diff
@@ -154,19 +156,35 @@ getBlocks
     -> [HeaderHash]
     -> d (Either GetBlocksError [Block])
 getBlocks logic enqueue nodeId tipHeader checkpoints = do
-    headers <- requestHeaders
-    blocks <- requestBlocks headers
-    return (Right (toList blocks))
-    -- TODO exception handling? Don't just catchj all.
+    -- It is apparently an error to request headers for the tipHeader and
+    -- [tipHeader], i.e. 1 checkpoint equal to the header of the block that
+    -- you want. Sure, it's a silly thing to do, but should it be an error?
+    --
+    -- Anyway, the procedure was and still is: if it's just one block you want,
+    -- then you can skip requesting the headers and go straight to requesting
+    -- the block itself.
+    blocks <- if singleBlockHeader
+              then requestBlocks (NewestFirst (pure tipHeader))
```

Comment:
This is incorrect. If we have only one tip and only one checkpoint it means that we have only first genesis block in our DB. We should act like `singleBlockHeader` is `False` here too.

Reply:
Calling this with one checkpoint and one tip does not imply anything about any DB.

Reply:
No it does. `singleBlockHeader` is defined as a function matching on checkpoints. If `singleBlockHeader` is true, then we have only one checkpoint in the request, which is only possible if we have first genesis block in DB.

Indeed, if we want to ask for one header only, we can skip asking for headers step. But this is different from code semantics here. `singleBlockHeader` will not ever be true if our tip is something different from first gblock. Moreover, checkpoints set, as defined by the only function that generates them, must contain log(k) elements if current blockchain complexity is >k.

Reply:
> we have only one checkpoint in the request, which is only possible if we have first genesis block in DB.

There's no inherent connection between a `MsgGetHeaders` value and a database.

Reply:
"checkpoints" is a network protocol term, function from current blockchain state, whatever. You may want to use any other set instead of checkpoints, but not in the main application logic, unless we modify the protocol. We're talking about datatype internal guarantees. Anyway, it's minor.

Reply:
"checkpoints"  is not a network protocol term. One does not need any network-related notions to define it. It's a set of hashes in some blockchain.

Reply:
Yes, sure, let's not argue about definitions. I'm trying to say just one thing here: if there's no valid usecase for passing some different checkpoints from what `getHeadersOlderExp` returns, we shouldn't expose it as interface, because otherwise somebody will mistakenly put something else in that argument. Alternatively, we can create a wrapper for it: `netwtype Checkpoints = Checkpoints (Set HeaderHash)`. There are internal invariants (at least one of these headers should be in common main chain over the last `k` headers, etc), so one should think twice before constructing a value manually.